Mathlib/RingTheory/GradedAlgebra/HomogeneousLocalization.lean

```diff
@@ -610,4 +614,98 @@ abbrev Away :=
 
 end
 
+section
+
+variable (𝒜)
+variable (B C : Type*) [CommRing B] [Algebra R B]
+variable (ℬ : ι → Submodule R B) [GradedAlgebra ℬ]
+variable (P : Submonoid A) (Q : Submonoid B)
+variable [CommRing C]
+
+
+/--
+Homogeneous localization as a subring
+-/
+def asSubring : Subring (Localization P) where
+  carrier :=
+  { a : at P | ∃ b : HomogeneousLocalization 𝒜 P, a = b.val }
+  mul_mem' := by
+    rintro _ _ ⟨a, rfl⟩ ⟨b, rfl⟩
+    exact ⟨a * b, mul_val a b |>.symm⟩
+  one_mem' := ⟨1, one_val |>.symm⟩
+  add_mem' := by
+    rintro _ _ ⟨a, rfl⟩ ⟨b, rfl⟩
+    exact ⟨a + b, add_val a b |>.symm⟩
+  zero_mem' := ⟨0, zero_val |>.symm⟩
+  neg_mem' := by
+    rintro _ ⟨a, rfl⟩
+    exact ⟨-a, neg_val a |>.symm⟩
+
+/--
+Homogeneous localization as a subring of normal localization
+-/
+def equivSubring : HomogeneousLocalization 𝒜 P ≃+* asSubring 𝒜 P where
+  toFun x := ⟨x.val, ⟨_, rfl⟩⟩
+  invFun x := .mk''
+    ⟨x.2.choose.deg,
+      ⟨x.2.choose.num, x.2.choose.num_mem_deg⟩,
+      ⟨x.2.choose.den, x.2.choose.den_mem_deg⟩,
+      x.2.choose.den_mem⟩
+  left_inv x := by
+    rw [ext_iff_val, val_mk'', x.eq_num_div_den, Localization.mk_eq_mk_iff,
+      Localization.r_iff_exists]
+    dsimp only
+    generalize_proofs h
+    have eq : x = h.choose := by
+      rw [ext_iff_val]
+      exact h.choose_spec
+    rw [← eq]
+    simp only [exists_const]
+  right_inv := by
+    rintro ⟨_, ⟨x, rfl⟩⟩
+    simp only [val_mk'', Subtype.mk.injEq]
+    generalize_proofs h1 h2
+    dsimp only at h1 h2
+    have eq : x = h1.choose := by
+      rw [ext_iff_val]
+      exact h1.choose_spec
+    conv_rhs => rw [x.eq_num_div_den, eq]
+  map_mul' _ _ := by
+    simp only [mul_val]
+    rfl
+  map_add' _ _ := by
+    simp only [add_val]
+    rfl
+
+variable {C P}
+/--
+If `g : A ⟶ C` is a ring homomorphism such that `g(P)` has only invertible elements,
+then we get a ring homomorphism `A⁰_P ⟶ C`
+-/
+def lift (g : A →+* C) (is_unit : ∀ x : P, IsUnit (g x)) :
+    HomogeneousLocalization 𝒜 P →+* C :=
+  RingHom.comp (IsLocalization.lift (M := P) (S := Localization P) is_unit) <|
+    algebraMap _ _
+
+/--
+If `g : A ⟶ B` is a ring homomorphism preserving degree such that `g(P) ⊆ Q`, then
+there is a ring homomorphism `A⁰_P ⟶ B⁰_Q`.
+-/
+def map (g : A →+* B)
+    (comap_le : P ≤ Q.comap g) (preserves_deg : ∀ a : A, ∀ i : ι, a ∈ 𝒜 i ↔ g a ∈ ℬ i) :
```

Comment:
I think you only need the => direction of `preserves_deg`?
Also can you add `map g _ _ (Quotient.mk'' x) = ?_` where `x : NumDenSameDeg 𝒜 P`?